### PR TITLE
Benny/redis

### PR DESCRIPTION
**Pull request by bennycio**
Changes:

- **Added** redis client to storage and funcs for setting and getting cached items
- **Changed** get unassigned and opensea get to have a `skip_cache` param that will forcibly not go search cache first
- **Changed** get unassigned and opensea get to check cache before getting and set cache after getting
- **Changed** storage funcs to be private because they are only used and only should be used in persist package

Considerations:

- What TTLs should we use for opensea get and get unassigned?

TODO:

- [x] Add to README how to start a local redis server with docker
- [ ] Test

### CODE REVIEW

File: persist/storage.go [hunk 1]
```diff
@@ -9,32 +9,61 @@ import (
...
-func NewMongoStorage(version int64, collName string, runtime *runtime.Runtime) *MongoStorage {
+func newStorage(version int64, collName string, runtime *runtime.Runtime) *storage {
 	coll := runtime.DB.MongoDB.Collection(collName)
-	return &MongoStorage{version: version, collection: coll}
+
+	return &storage{version: version, collection: coll}
+}
+
+// withRedis attaches a redis client to the Storage instance
+func (m *storage) withRedis(rdb redisDB, runtime *runtime.Runtime) *storage {
```
Comment (Robinnnnn) [inline persist/storage.go:54]: 😍

File: runtime/config.go [hunk 2]
```diff
@@ -14,10 +14,13 @@ const (
...
-	mongoTLSSecretName    = "GLRY_TLS"
-	mongoUseTLS           = "GLRY_MONGO_USE_TLS"
-	mongoDBname           = "GLRY_MONGO_DB_NAME"
+	mongoURLSecretName = "projects/1066359838176/secrets/GLRY_MONGO_URL/versions/latest"
+	mongoTLSSecretName = "projects/1066359838176/secrets/GLRY_TLS/versions/latest"
+	mongoUseTLS        = "GLRY_MONGO_USE_TLS"
+	mongoDBname        = "GLRY_MONGO_DB_NAME"
+
+	redisURL            = "REDIS_URL"
+	redisPassSecretName = "projects/1066359838176/secrets/GLRY_REDIS_PASS/versions/latest"
```
Comment (Robinnnnn) [inline runtime/config.go:23]: wow, thank you for setting this up already!

File: README.md [hunk 3]
```diff
@@ -27,12 +27,16 @@ $ ./bin/main
...
 
-The app will connect to a local mongo instance by default. To spin it up, you can use the official docker container:
+The app will connect to a local mongo instance and local redis instance by default. To spin it up, you can use the official docker containers:
 
 ```bash
 $ docker run -p 27017:27017 mongo
 ```
 
+```bash
+$ docker run -p 6379:6379 redis
```
Comment (kaitoo1) [inline README.md:37]: nice, I added this command to our Backend Onboarding doc in notion

File: persist/storage.go [hunk 4]
```diff
@@ -10,31 +10,60 @@ import (
...
+const (
+	// CollectionsUnassignedRDB is a throttled cache for expensive queries finding unassigned NFTs
+	CollectionsUnassignedRDB redisDB = iota
+	// OpenseaGetRDB is a throttled cache for expensive queries finding Opensea NFTs
+	OpenseaGetRDB
+)
+
+var (
+	collectionUnassignedTTL time.Duration = time.Minute * 1
+	opeseaGetTTL            time.Duration = time.Minute * 1
```
Comment (kaitoo1) [inline persist/storage.go:33]:
```suggestion
	openseaGetTTL            time.Duration = time.Minute * 1
```

File: persist/nft.go [hunk 5]
```diff
@@ -211,6 +217,37 @@ func NftRemoveDifference(pCtx context.Context, pNfts []*Nft, pWalletAddress stri
...
+func NftOpenseaCacheSet(pCtx context.Context, pWalletAddress string, pNfts []*Nft, pRuntime *runtime.Runtime) error {
+
+	mp := newStorage(0, nftColName, pRuntime).withRedis(OpenseaGetRDB, pRuntime)
+	defer mp.cacheClose()
+
+	toCache, err := json.Marshal(pNfts)
+	if err != nil {
+		return err
+	}
+	return mp.cacheSet(pCtx, string(pWalletAddress), toCache, opeseaGetTTL)
```
Comment (kaitoo1) [inline persist/nft.go:230]:
```suggestion
	return mp.cacheSet(pCtx, string(pWalletAddress), toCache, openseaGetTTL)
```

File: persist/nft.go [hunk 6]
```diff
@@ -211,6 +217,37 @@ func NftRemoveDifference(pCtx context.Context, pNfts []*Nft, pWalletAddress stri
...
+}
+
+// NftOpenseaCacheGet adds a set of nfts to the opensea cache under a given wallet address
```
Comment (kaitoo1) [inline persist/nft.go:233]: Should this comment say something like `.. gets a set of nfts from the opensea cache..` instead?

File: persist/storage.go [hunk 7]
```diff
@@ -10,31 +10,60 @@ import (
...
 
+	"github.com/go-redis/redis/v8"
 	"github.com/mikeydub/go-gallery/copy"
 	"github.com/mikeydub/go-gallery/runtime"
 	"go.mongodb.org/mongo-driver/bson"
 	"go.mongodb.org/mongo-driver/mongo"
 	"go.mongodb.org/mongo-driver/mongo/options"
 )
 
+// RedisDB represents the database number to use for the redis client
```
Comment (kaitoo1) [inline persist/storage.go:21]: thanks for the helpful comments!

**Review by kaitoo1 — APPROVED** [review]
Looks great! thanks for adding this.
I left a couple of minor comments.
Also are there any tests we can add to verify that the skipCache param works as intended?

**Comment by kaitoo1** [comment]
For TTL, what about starting with an hour? And we can see if we want it to be shorter or longer

**Comment by kaitoo1** [comment]
We might want the TTL for getting unassigned NFTs to be shorter though.. 5-10 min?

File: persist/nft.go (hunk 6 above)
Comment (bennycio) [reply persist/nft.go:233]: Yes it sure should! Will make this change!

**Comment by kaitoo1** [comment]
thanks for addressing comments! LGTM

**Pull request merged**